packages/cli/src/commands/hydrogen/upgrade.ts

```diff
@@ -0,0 +1,2931 @@
+  static description = 'Upgrade Remix and Hydrogen npm dependencies.';
+
+  static flags = {
+    path: commonFlags.path,
+    version: Flags.string({
+      description: 'A target hydrogen version to update to',
+      required: false,
+      char: 'v',
+    }),
+    ['dry-run']: Flags.boolean({
+      description:
+        'Generate a summary and .md file with upgrade instructions without actually upgrading the dependencies',
+      required: false,
+      default: false,
+      char: 'd',
+    }),
+  };
+
+  async run(): Promise<void> {
+    const {flags} = await this.parse(Upgrade);
+    const appPath = flags.path ? path.resolve(flags.path) : process.cwd();
+    const version = flags.version;
+    const dryRun = Boolean(flags['dry-run']);
+    await runUpgrade({appPath, version, dryRun});
+  }
+}
+
+let CACHED_CHANGELOG: ChangeLog | null = null;
+
+export async function runUpgrade({
+  appPath,
+  dryRun,
+  version: targetVersion,
+}: {
+  appPath: string;
+  dryRun: boolean;
+  version?: string;
+}) {
+  await checkIsGitRepo(appPath);
+
+  await checkDirtyGitBranch(appPath);
+
+  const current = await getHydrogenVersion({appPath});
+
+  if (!current?.currentVersion) {
+    throw new Error('Failed to get current Hydrogen version');
+  }
+
+  const {currentVersion, currentDependencies} = current;
+
+  // TODO: swap when we merge to production
+  // const snapshots = await fetchChangelog();
+  const changelog = await fetchTempChangelog();
+
+  const {availableUpgrades} = getAvailableUpgrades({
+    releases: changelog.releases,
+    currentVersion,
+  });
+
+  if (!availableUpgrades?.length) {
+    const upToDateMessage = `You are on the latest Hydrogen version: ${getAbsoluteVersion(
+      currentVersion,
+    )}`;
+
+    renderSuccess({headline: upToDateMessage});
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+
+    return;
+  }
+
+  // Prompt the user to select a version from the list of available upgrades
+  const selectedRelease = await getSelectedRelease({
+    currentVersion,
+    targetVersion,
+    availableUpgrades,
+  });
+
+  if (!selectedRelease) {
+    renderFatalError({
+      name: 'error',
+      type: 0,
+      message: 'No Hydrogen version selected',
+      tryMessage: `Please try again later`,
+    });
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+
+    return;
+  }
+
+  // Get an aggregate list of features and fixes included in the upgrade versions range
+  const cumulativeRelease = getCummulativeRelease({
+    releases: changelog.releases,
+    currentVersion,
+    selectedRelease,
+  });
+
+  // Prompt the user to confirm the upgrade and display a list of features and fixes
+  const confirmed = await displayConfirmation({
+    appPath,
+    cumulativeRelease,
+    dryRun,
+    selectedRelease,
+    targetVersion,
+  });
+
+  // Exit if the user chose to return to the version selection prompt
+  if (!confirmed) {
+    return;
+  }
```

Comment:
I see `displayConfirmation` is recursively calling `runUpgrade` to allow the user change selection, which is a nice feature.

Can we make this iteratively instead of recursively? That way we can skip re-computing things like checking git directory, downloading changelog, finding releases, etc:

```js
let confirmed = false;

while (!confirmed) {
  const selectedRelease = await selectRelease(...)
  if (!selectedRelease) throw new AbortError(...)

  const cumulativeRelease = getCummulativeRelease(...)
  confirmed = await displayConfirmation(...)
}
```